base/arraymath.jl

```diff
@@ -58,69 +58,83 @@ end
-    end
-    if nnd==nd
-        # reverse along the only non-singleton dimension
-        for i = 1:sd
-            B[i] = A[sd+1-i]
-        end
-        return B
-    end
+# Examples
+```jldoctest
+julia> b = Int64[1 2; 3 4]
+2×2 Matrix{Int64}:
+ 1  2
+ 3  4
+
+julia> reverse(b, dims=2)
+2×2 Matrix{Int64}:
+ 2  1
+ 4  3
+
+ julia> reverse(b)
+2×2 Array{Int64,2}:
+ 4  3
+ 2  1
+```
+
+!!! compat "Julia 1.6"
+    Prior to Julia 1.6, only single-integer `dims` are supported in `reverse`.
+"""
+reverse(A::AbstractArray; dims=:) = _reverse(A, dims)
+_reverse(A, dims) = reverse!(copymutable(A); dims)
+
+"""
+    reverse!(A; dims=:)
+
+Like [`reverse`](@ref), but operates in-place in `A`.
 
-    d_in = size(A)
-    M = 1
-    for i = 1:d-1
-        M *= d_in[i]
+!!! compat "Julia 1.6"
+    Multidimensional `reverse!` requires Julia 1.6.
+"""
+reverse!(A::AbstractArray; dims=:) = _reverse!(A, dims)
+_reverse!(A::AbstractArray{<:Any,N}, ::Colon) where {N} = _reverse!(A, ntuple(identity, Val{N}()))
+_reverse!(A, dim::Integer) = _reverse!(A, (Int(dim),))
+_reverse!(A, dims::NTuple{M,Integer}) where {M} = _reverse!(A, Int.(dims))
+function _reverse!(A::AbstractArray{<:Any,N}, dims::NTuple{M,Int}) where {N,M}
+    if N < M || !allunique(dims) || !all(d -> 1 ≤ d ≤ N, dims)
+        throw(ArgumentError("invalid dimensions $dims in reverse!"))
     end
-    N = length(A)
-    stride = M * sd
-
-    if M==1
-        for j = 0:stride:(N-stride)
-            for i = 1:sd
-                ri = sd+1-i
-                B[j + ri] = A[j + i]
-            end
-        end
-    else
-        if allocatedinline(T) && M>200
-            for i = 1:sd
-                ri = sd+1-i
-                for j=0:stride:(N-stride)
-                    offs = j + 1 + (i-1)*M
-                    boffs = j + 1 + (ri-1)*M
-                    copyto!(B, boffs, A, offs, M)
-                end
-            end
-        else
-            for i = 1:sd
-                ri = sd+1-i
-                for j=0:stride:(N-stride)
-                    offs = j + 1 + (i-1)*M
-                    boffs = j + 1 + (ri-1)*M
-                    for k=0:(M-1)
-                        B[boffs + k] = A[offs + k]
-                    end
-                end
-            end
+    M == 0 && return A # nothing to reverse
+
+    idx = Vector{Int}(undef, N) # temporary array for index calculations
+
+    # compute sizes for half of reversed array
+    ntuple(k -> @inbounds(idx[k] = size(A,k)), Val{N}())
+    @inbounds idx[dims[1]] = idx[dims[1]] >> 1
+
+    last1 = ntuple(k -> lastindex(A,k)+firstindex(A,k), Val{N}())
+    for i in CartesianIndices(ntuple(k -> firstindex(A,k):firstindex(A,k)-1+@inbounds(idx[k]), Val{N}()))
+        ntuple(k -> @inbounds(idx[k] = i[k]), Val{N}())
+        ntuple(Val{M}()) do k
+            @inbounds j = dims[k]
+            @inbounds idx[j] = last1[j] - idx[j]
         end
+        iᵣ = CartesianIndex(ntuple(k -> @inbounds(idx[k]), Val{N}()))
```

Comment:
Would it be worthwhile to set up an `NTuple{N,Bool}` or `Vector{Bool}` of length `N`, say `dimrev`, that encodes for every dimension of `A` whether it should be reversed? Then this would become
```julia
CartesianIndex(ntuple(k -> @inbounds(dimrev[k] ? last1[k] - i[k] : i[k]), Val{N}()))
```
and the two `ntuple` constructions above could be removed (at the expense of a similar construct before the loop to set up `dimrev`, of course).

I think to me, that would make this a bit more readable, but that may just be my personal taste, so please feel free to leave it as is if you prefer.

Reply:
Yes, that's reasonable.  Might even be faster (maybe with `ifelse`).

Reply:
One could even do `ifelse.(dimrev, last1 .- i, i)`, since tuple broadcasts are inlined/unrolled.

Reply:
Yup, the `ifelse` approach is indeed faster as well as simpler, and it lets me cut down the number of allocations to 0 for `reverse!`

Reply:
Yes, much nicer this way.